[WIP] Windows support

lukehoban opened the pull request:
Initial work on porting Delve to support Windows.  

There's a lot more to do before this is ready to be merged, but most of the core infrastructure for Windows support is in place, and some basic Delve scenarios are working in simple scenarios (launching, setting breakpoints, continuing, breaking on breakpoints, inspecting registers, seeing sources, inspecting locals, exiting, etc.).

Fixes #198.

ignatov commented:
Great!

lukehoban commented:
A quick update on progress:

- All debugger features are generally working (except `attach`).
- Unit tests are clean on Windows.
- Unit tests remain clean on Mac/Linux (modulo known flakiness).
- Appveyor CI support added to guard against regressions.

A few of the known remaining work items:

- The `TestNextNetHTTP` test is disabled and needs to be investigated.
- Attach support is not yet enabled on Windows.
- Thread information is not displayed nicely for most non-active threads because the PC on those threads is in `ntdll` or other dynamically loaded code.  To address this, we would need to load symbols on Windows and merge that information with the DWARF symbols.  This should be possible, but worth discussing before implementing.
- There are a handful of TODOs to investigate and improve on workarounds in the existing implementation.
- Effort was made to have minimal impact on the Mac/Linux codebase so far.  In the future, it likely will make sense to define a more formal platform abstraction layer that is platform independent.

However, at this point it is worth starting to get more eyes on this code and discussing if/when this should be merged into master.

/cc @derekparker

aarzilli commented on proc/threads_windows.go:
```diff
@@ -0,0 +1,142 @@
+	if err != nil {
+		return err
+	}
+	_, err = t.dbp.trapWait(0)
```
I think this is a bug, calling `trapWait` inside `singleStep` means delve will get stuck on breakpoints set on 1 byte instructions. See #262 and the commit I added today to #259.

derekparker replied:
That's true. @aarzilli is there a test for that? If not we should add one, because for now let's simply not use trapWait here, but shortly after this and some other PRs land I'd like to look into ensuring that trapWait is safe to call in that situation. Either way I'd like to prevent that as a regression in the future.

aarzilli replied:
#259 has `TestIssue262` which does catch this on linux and OS X, however it's highly dependent on what the compiler does.

derekparker commented on proc/proc_windows.go:
```diff
@@ -0,0 +1,369 @@
+		var err error
+		var tid int
+		var exitCode int
+		dbp.execPtraceFunc(func() {
```
Is there a restriction in Windows that these syscalls must originate from the same thread?

lukehoban replied:
That's right - both WaitForDebugEvent and ContinueDebugEvent must be called from the same thread that created the process.

Details here (https://msdn.microsoft.com/en-us/library/windows/desktop/ms679285(v=vs.85).aspx) and here (https://msdn.microsoft.com/en-us/library/windows/desktop/ms681423(v=vs.85).aspx).

derekparker commented on proc/test/support.go:
```diff
@@ -48,13 +50,23 @@ func BuildFixture(name string) Fixture {
 	}
 
 	source, _ := filepath.Abs(path)
+	if runtime.GOOS == "windows" {
```
Same question from above applies here.

lukehoban replied:
Same answer as above - happy to change to use the code without a platform test if that's preferred.

derekparker commented on proc/threads_windows.go:
```diff
@@ -0,0 +1,142 @@
+	//context.Rip--;
+	context.EFlags |= 0x100;
+	
+	
```
nit: cleanup extra newline here.

derekparker commented on proc/threads_windows.go:
```diff
@@ -0,0 +1,142 @@
+	t.running = true
+	var res C.WINBOOL
+	t.dbp.execPtraceFunc(func() {
+		//TODO: Note that we are ignoring the thread we were asked to continue and are continuing the 
```
Any reason for this? A `thread.resume` call should only act on that that.

lukehoban replied:
This is a temporary workaround for a problem somewhere else that is causing requests for resuming on a thread that should not be resumed. 

I believe it is related to the issue with the resume call from singleStep.  We should be able to remove this workaround once those other places are fixed.

lukehoban replied:
One other note is that the Windows debugging APIs only allow continuing the same thread on which the previous debug event was raised.  When the debug event is raised all threads are paused and then when ContiuneDebugEvent is called all threads resume.

To correctly singleStep (or continue only a single thread while keeping others stopped), we will need to pause all the threads other than the one we want to step, then continue, then resume all the threads again after the break event is raised for the step.

derekparker commented on proc/threads_windows.go:
```diff
@@ -0,0 +1,142 @@
+}
+
+func (thread *Thread) blocked() bool {
+	// TODO: Probably incorrect - what are teh runtime functions that
```
I can look into this.

derekparker commented on service/debugger/locations.go:
```diff
@@ -111,6 +116,10 @@ func parseLocationSpecDefault(locStr, rest string) (LocationSpec, error) {
 	spec.Base = v[0]
+	if runtime.GOOS == "windows" {
+		// Normalize paths to "/" to match Go symbol table paths
+		spec.Base = strings.Replace(spec.Base, "\\", "/", -1)
```
See above filepath comment.

lukehoban commented on proc/proc.go:
```diff
@@ -151,6 +149,9 @@ func (dbp *Process) LoadInformation(path string) error {
 }
 
 func (dbp *Process) FindFileLocation(fileName string, lineno int) (uint64, error) {
+	if runtime.GOOS == "windows" {
```
Unfortunately no - this is doing the opposite of that library function.  We need to turn a platform-specific path into one which always uses '/', since for some reason that is how paths are being encoded into DWARF, even on Windows.  

We could avoid the runtime check by always doing something like this - which will just be a no-op on unix:

```go
fileName = strings.Replace(fileName, string(os.PathSeparator), "/", -1)
```

lukehoban replied:
I take it back - `filePath.ToSlash` does what's needed here and in the other two places you called out.  I'll use that.

lukehoban commented on proc/proc_windows.go:
```diff
@@ -0,0 +1,369 @@
+		// TODO - We're ignoring the results because we assume we'll immediately hit
+		// the default breakpoint that Windows sets at process creation.
+		// Should perhaps be testing that we're not overlooking an exit event or similar?
+		_, _, err = dbp.waitForDebugEvent()
```
I'm not certain it's even possible, but I've left the TODO here and added a note about returning `ProcessExitedError` if it is possible.

derekparker commented:
@lukehoban sorry for the delay, have gotten really busy lately. I will be taking another look tomorrow. How is the work going to finish implementing features and get all tests to pass?

derekparker commented:
@lukehoban do you plan on 100% parity with this PR or to get an initial implementation? Because I'm fine with merging a solid implementation that doesn't have 100% parity, as long as we have issues to track the other features that must be implemented.

aarzilli commented:
My note about the `trapWait` call can be disregarded given the `Continue` refactoring we did on the other branch.

lukehoban commented:
@derekparker Apologies for the delay getting back to you.

I do think this is close to ready to merge in with a discrete set of known issues we can track in the issue tracker.

There is only one truly "incorrect" thing I am aware of in the current implementation, which is the issue mentioned in https://github.com/derekparker/delve/pull/276#discussion_r45100529.  

Other than that, the remaining points from https://github.com/derekparker/delve/pull/276#issuecomment-155274585 are places where there is not yet 100% parity, but what does work works correctly.  I'm happy to work on these as issues after merging this PR.

Certainly, merging this in with the CI turned on will help to make sure the Windows support doesn't regress going forward.

On the singleStep issue, I'll try to find some time to look at that in the next few days, and to bring this branch up to date with master.  Once I have an update on that, we can decide whether to merge this in.

dprime commented:
Hey,

I'm just trying to build this to test it out on my windows setup. Having a few problems getting things running. Should I be building on windows or cross compiling from linux to windows? When attempting to build on windows, it moans about lacking a gcc install and when on linux, it complains about `../../service/api/conversions.go:9:2: C source files not allowed when not using cgo or SWIG: threads_windows.c`
Any thoughts?

lukehoban commented:
Yeah - Delve uses cgo heavily, and cgo on Windows [requires a GCC toolchain](https://github.com/golang/go/wiki/cgo#windows).  I've been compiling on Windows with the MinGW64 tools.  

The [appveyor.yml](https://github.com/lukehoban/delve/blob/windows/appveyor.yml) has one particular toolchain configuration that works and is being used in the CI toolchain: `x86_64-4.9.2-release-win32-seh-rt_v4-rev3`.

Locally I've used a slightly different MinGW version: `x86_64-5.2.0-posix-seh-rt_v4-rev0`.

dprime commented:
Thanks, giving that a spin now.

dprime commented:
Builds and works very nicely. Issue now is how to integrate into the windows version of intellij? It just works in linux and I'm trying to figure out how to coax the windows version to realise I have a working delve binary. Sorry to hijack this PR thread for this.

ignatov commented:
@dprime You should create a small patch and build the plugin from sources.
E.g. `com.goide.dlv.DlvDebugProcess#IS_DLV_DISABLED = SystemInfo.is32Bit;`
and setup a desired `customDlvPath` in `gradle.properties`.

mistbow-git commented:
Confirmed! Windows version builds and works perfectly!
I need debugging support on windows very much!
Please resolve all conflicts and merge with main branch!
And build idea plugin with support dalve for windows already...
Thanks!

zanven42 commented:
firstly i would like to say thank you for everyone who has been working on windows support and i am super excited to start using it, 
Secondly @mistbow-git these people are doing this for free without being paid (most likely) if you really want to see it in and finished and you are not willing to wait till they have the spare time to do it i suggest you start cracking on finishing it off yourself and give them a hand.

mistbow-git commented:
@zanven42 Join the words of gratitude!
I builded dalve for windows last nite from souces https://github.com/mistbow-git/delve with last commits from @derekparker master. All working good! But one test failed (
Then I build a custom plugin for IDEA and testing debugging feature. Its working!
Thank you very much for all! Waiting for relise!

zanven42 commented:
yeah after i posted i noticed you were doing things and it got me excited hehe. ill have to get it building tonight and have a look.

kolkov commented:
If you want, I can share files ready for testing...

zanven42 commented:
that would be handy thanks.

kolkov commented:
http://kolkov.ru/share/dlv/dlv.rar 
dlv.exe folder d:/Go/bin

derekparker commented:
@lukehoban can you please rebase this on master, squash commits and push. Could you also open issues for any functionality that this PR does not provide? I'd like to get this branch merged in soon, and add more visibility into what hasn't been finished.

Thanks again for all your work on this port!

shiena commented:
I also tried the windows version.
There is almost no problem, but there is one question.
Does not display the string in `fmt.Println("hello")` In the Windows version.
Do I have failed anything?

![dlv2](https://cloud.githubusercontent.com/assets/1295639/12234973/e9a9434e-b8b3-11e5-9ddb-08717f6465b3.png)

lukehoban commented:
@derekparker Looks like there's some work to do to update this to catch up to the refactorings/changes that were merged into master in the last week or two.  I'm working on that now.

lukehoban commented:
@shiena Yes - I believe that's a missing feature of the Windows port currently.  I'll make sure that's one of the issues we open before merging this.

aarzilli commented:
@lukehoban the work that was done in `handleBreakpointOnThread` is now inside `setCurrentBreakpoints` (which should find all threads that are stopped on a breakpoint if there is more than one).

The process is always restarted by calling `(*Process).resume()`, this method is responsible for making threads stopped on breakpoints step over their breakpoints correctly, from what I understand this should be a better fit for windows.

shiena commented:
@lukehoban Thanks for your reply. I see.
This patch is sufficiently great even while now.

lukehoban commented:
I've pushed a few new commits which:

1. Brings this branch up to date with `derekparker/master`
2. Addresses some latent bugs that @aarzilli's recent refactorings uncovered
3. Fixes the issue with single stepping potentially stepping the wrong thread (https://github.com/derekparker/delve/pull/276#discussion_r45100529)
4. Adds support for redirecting stdin/stdout/stderr from the debuggee per @shiena's feedback

This appears to be passing tests on both Travis and Appveyor.

I will pull together the list of known issues and cleanup items and will squash these commits down tomorrow.

derekparker commented:
@lukehoban thank you for your work on this. I'll be looking through the code soon for a final review and we will get this merged in.

lukehoban commented:
@derekparker Great - thanks.  I'm working on some cleanup and the list of open issues, which I'll have available soon.  Might be worth waiting on that before doing another review pass.

lukehoban commented:
This branch is now up to date with master and squashed.  

Tests are passing on Appveyor (https://ci.appveyor.com/project/lukehoban/delve) and Travis (https://travis-ci.org/derekparker/delve/builds/103290576).

Code reviews welcomed. @derekparker @aarzilli 

Here's an inventory of remaining work I'm aware of.  I'm okay with merging this branch and opening issues to track each of these.  If there are any that should be fixed before merging, let me know.

- [ ] **TestDebugLinePrologueParser**
Only on some  machines and Go versions.  Appears to be because of buffers in the debugline data which are larger than they should be in Windows binaries?  I suspect there is a bug in the Go compiler on Windows that we'll need to work around.

- [ ] **TestNextConcurrent flakiness**
Occasional occurrences of the following errors:
`proc_test.go:355: Program did not continue to correct next location expected 11 was parallel_next.go:10`
`proc_test.go:355: Program did not continue to correct next location expected 11 was parallel_next.go:8`
`proc_test.go:355: Program did not continue to correct next location expected 10 was parallel_next.go:8`
`proc_test.go:410: Program did not continue to correct next location expected 11 was parallel_next.go:10`
`proc_test.go:403: Initial breakpoint triggered twice for the same goroutine`

- [ ] **Attach is not supported**
Should also add test coverage for attach on all platforms.

- [ ] **TestNextNetHTTP is disabled on Windows**
Temporarily disabled this test, need to investigate the failure.

- [ ] **Should we load symbols on Windows?**
Thread information is not displayed nicely for most non-active threads because the PC on those threads is in ntdll or other dynamically loaded code. To address this, we would need to load symbols on Windows and merge that information with the DWARF symbols. This should be possible, but worth discussing before implementing.

- [ ] **thread.halt() may need to actually halt threads in some cases**
See comment in implementation.

- [ ] **Can we remove canContinue?**
dbp.Step currently relies on checking this, but would be nice to factor it to not need this concept in addition to blocked and stopped.

- [ ] **Infrequent failures in TestBreakPointCount**
Pretty infrequent occurence, and seemingly always with 101 + 99.  So it appears that one of the breakpoint hits is getting recorded as being on the wrong thread. 

Once merged, we will also want to do:

- [ ] **Add Appveyor project tracking derekparker/delve and badge in README**

- [ ] **Update docs with info on building/running on Windows**
Mostly just MinGW requirement which is likely not present for many Windows Go users.

aarzilli commented:
> Can we remove canContinue?
> dbp.Step currently relies on checking this, but would be nice to factor it to not need this concept in addition to blocked and stopped.

Are you talking about `canStep`? In that case, my opinion is that `proc.(*Process).Step` shouldn't step all threads but just the one associated with `SelectedGoroutine` (if there is one). I don't think a user would expect all threads to step and I'm not sure what use that would have.

lukehoban commented:
Re: @aarzilli comment on `dbp.resume()`:

```go
	// In case we are now on a different thread, make sure we resume
	// the thread that is broken.
	thread = dbp.Threads[dbp.os.breakThread]
	if err := thread.resume(); err != nil {
		return err
	}
```
> Under what circumstance does this happen? Could it be that when two breakpoints are hit simultaneously the debug event for the second breakpoint gets erroneusly ignored here? This could be an explanation for some of the test failures.

The underlying problem is in `singleStep`.  The assumption there is that if we set the processor trap flag on the current thread, then suspend all the threads other than the current thread, then `ContinueDebugEvent`/`WaitForDebugEvent`, that the debug event we'll get is the single step, and therefore will be on the same thread we were on prior to the call to `step`.

It appears though that in rare cases, the debug event returned for `WaitForDebugEvent` is on a different thread.  The only way I can explain this is that either `SuspendThread` is not actually suspending the thread, or somehow the additional debug event was queued concurrently with the event raised from `WaitForDebugEvent` so that it gets returned before the single step.  

The second of these I am nearly certain should not be possible.  This would be along the lines of your suggestion that perhaps the second breakpoint event is being raised here before the `singleStep` gets returned.  But with the way the Windows debug APIs work, all threads should be synchronized on debug events, with every thread being suspended when a debug event is raised.  So there really should be no notion of two breakpoints being hit simultaneously.  (Though two threads may perhaps be able to have their PC on the breakpoint, with the second ready to raise another breakpoint exception debugevent once it's thread is continued?).  If this is the issue though, it's going to be hard to fix, since there would never be a safe way to singleStep without potentially triggering other debug events that should reenter other parts of Delve's implementation.

The former seems more likely, and [this post](https://blogs.msdn.microsoft.com/oldnewthing/20150205-00/?p=44743) suggests this may be where the problem is, and a potential workaround (calling `GetContext` to force synchronizing the effect of the `SuspendThread` call).

From a brief investigation just now, it seems plausible that all of the other intermittent test failures occur only after this condition occurs (though it happens a little more frequently than actual test failures).  In the last 20 minutes of running tests, in every case where there has been a test failure, this unexpected `singleStep` condition was hit at least once during the test.

I'll look into this further...

lukehoban commented:
> my opinion is that proc.(*Process).Step shouldn't step all threads but just the one associated with SelectedGoroutine (if there is one). I don't think a user would expect all threads to step and I'm not sure what use that would have.

Yes, that was what I would have expected as well.  I wasn't sure if there was some Linux/OSX specific reason for expecting all threads to step instead of just the SelectedGoroutine's thread.

I quickly tested on OSX changing `proc.(*Process).Step` to only step the SelectedGoroutine's thread, and I don't see any test failures.  Is there a scenario or test case that highlights the need for the current logic?  Or should this be changed across all platforms?

> Other than the go fmt nitpick this looks good to me.

I've fixed the `go fmt` in `threads_windows.go`.

joelday commented:
Is it reasonable to conclude that this is far enough, but with known issues?

joelday commented:
Also, @lukehoban, instructions on how to install from your PR/branch?

zanven42 commented:
have mingw installed (64 if u dont know if ur go is 32bit or 64bit install). 

in derekparker's repo add lukehoban as a extra remote and checkout he's windows branch. 

go to the folder cmd -> dlv. open a command prompt. go build

hope that helps.

joelday commented:
@zanven42 Thanks! Managed to figure that out a few minutes ago.

This is what worked for me:
> git clone -q --branch=windows https://github.com/lukehoban/delve.git %GOPATH%\src\github.com\derekparker\delve
> set GO15VENDOREXPERIMENT=1
> cd %GOPATH%\src\github.com\derekparker\delve
> mingw32-make install

kolkov commented:
I use go to build dlv.exe
cd D:\Go\src\github.com\derekparker\delve\cmd\dlv
and then run: go build.

derekparker commented on proc/proc_windows.go:
```diff
@@ -0,0 +1,423 @@
+		return nil, err
+	}
+
+    argv0, _ := syscall.UTF16PtrFromString(argv0Go)
```
Looks like there is some strange formatting in this function. Can you make sure gofmt has been run on all source files?

derekparker commented:
> Yes, that was what I would have expected as well. I wasn't sure if there was some Linux/OSX specific reason for expecting all threads to step instead of just the SelectedGoroutine's thread.
>
> I quickly tested on OSX changing proc.(*Process).Step to only step the SelectedGoroutine's thread, and I don't see any test failures. Is there a scenario or test case that highlights the need for the current logic? Or should this be changed across all platforms?

Per discussion on IRC with @aarzilli we have decided that Step should only step the "current" thread. The concerns I had in mind when I initially implemented this to step all threads is essentially a non-issue, and is causing more problems.

I'm going to submit a patch today to step only the "current" thread, as well as do a final review of this branch. Once that patch lands @lukehoban the last major change to this branch (aside from review comment fixes) should be to match the new behavior of step. After all that is finished this branch will be merged and outstanding issues will have GH issues created to track them. The plan is (depending on how quickly the changes can be made) to have this branch merged into master this week (honestly, the ideal situation would be this evening or tomorrow). From there we can address the outstanding issues individually.

derekparker commented on proc/proc_windows.go:
```diff
@@ -0,0 +1,423 @@
+
+// Attach to an existing process with the given PID.
+func Attach(pid int) (*Process, error) {
+	fmt.Println("Attach")
```
Let's remove this and just return the error.

derekparker commented on proc/proc_windows.go:
```diff
@@ -0,0 +1,423 @@
+}
+
+// Kill kills the process.
+func (dbp *Process) Kill() (err error) {
```
Named return isn't being used, remove it in favor of explicit return arg.

derekparker commented on proc/proc_windows.go:
```diff
@@ -0,0 +1,423 @@
+	return
+}
+
+func (dbp *Process) requestManualStop() (err error) {
```
Same comment as above re: named return.

derekparker commented on proc/proc_windows.go:
```diff
@@ -0,0 +1,423 @@
+func (dbp *Process) requestManualStop() (err error) {
+	res := C.DebugBreakProcess(C.HANDLE(dbp.os.hProcess))
+	if res == C.FALSE {
+		return fmt.Errorf("Failed to break process %d", dbp.Pid)
```
nit: Lowercase `Failed`.

derekparker commented on proc/proc_windows.go:
```diff
@@ -0,0 +1,423 @@
+		// Wait for a debug event...
+		res := C.WaitForDebugEvent(&debugEvent, C.INFINITE)
+		if res == C.WINBOOL(0) {
+			return 0, 0, fmt.Errorf("Could not WaitForDebugEvent")
```
nit: lowercase `Could`.

derekparker commented on proc/proc_windows.go:
```diff
@@ -0,0 +1,423 @@
+				return 0, 0, err
+			}
+			break
+		case C.LOAD_DLL_DEBUG_EVENT, C.UNLOAD_DLL_DEBUG_EVENT, C.EXIT_THREAD_DEBUG_EVENT, C.OUTPUT_DEBUG_STRING_EVENT, C.RIP_EVENT:
```
Why leave a thread exiting unimplemented? Should be pretty simple, just remove it from the list of known threads. That way dead threads won't clutter output of `threads` command.

derekparker commented on proc/proc_windows.go:
```diff
@@ -0,0 +1,423 @@
+}
+
+func (dbp *Process) wait(pid, options int) (int, *sys.WaitStatus, error) {
+	fmt.Println("wait")
```
Remove this Println, just return error.

derekparker commented on proc/proc_windows.go:
```diff
@@ -0,0 +1,423 @@
+
+func (dbp *Process) wait(pid, options int) (int, *sys.WaitStatus, error) {
+	fmt.Println("wait")
+	return 0, nil, fmt.Errorf("Not implemented: wait")
```
nit: lowercase `Not`.

ignatov commented:
@lukehoban Could you ping me directly when the PR will be merged, please.

derekparker commented on proc/proc_windows.go:
```diff
@@ -0,0 +1,423 @@
+	// in this case, one for each thread, so we should only handle the BP hit
+	// on the thread that the debugger was evented on.
+
+	err := trapthread.SetCurrentBreakpoint()
```
nit: rewrite as `return trapthread.SetCurrentBreakpoint()`

derekparker commented on proc/ptrace_windows.go:
```diff
@@ -0,0 +1,13 @@
+)
+
+func PtraceAttach(pid int) error {
+	return fmt.Errorf("Not implemented: PtraceAttach")
```
nit: lowercase `Not`.

derekparker commented on proc/ptrace_windows.go:
```diff
@@ -0,0 +1,13 @@
+}
+
+func PtraceDetach(tid, sig int) error {
+	return fmt.Errorf("Not implemented: PtraceDetach")
```
nit: lowercase `Not`.

derekparker commented on proc/registers_windows_amd64.go:
```diff
@@ -0,0 +1,166 @@
+}
+
+// SetPC sets the RIP register to the value specified by `pc`.
+func (r *Regs) SetPC(thread *Thread, pc uint64) error {
```
General nit: lowercase first letter of all new error messages.

derekparker commented on proc/registers_windows_amd64.go:
```diff
@@ -0,0 +1,166 @@
+	return regs, nil
+}
+
+func (thread *Thread) saveRegisters() (Registers, error) {
```
Remove Println's and just return error.

derekparker commented on proc/threads_windows.c:
```diff
@@ -0,0 +1,16 @@
+
+typedef NTSTATUS (WINAPI *pNtQIT)(HANDLE, LONG, PVOID, ULONG, PULONG);
+
+WINBOOL thread_basic_information(HANDLE h, THREAD_BASIC_INFORMATION* addr) {
```
Looks like some formatting issues here.

derekparker commented on proc/threads_windows.go:
```diff
@@ -0,0 +1,171 @@
+}
+
+func (t *Thread) blocked() bool {
+	// TODO: Probably incorrect - what are teh runtime functions that
```
s/teh/the/

derekparker commented:
@lukehoban https://github.com/derekparker/delve/pull/351 has landed, fixing the single step behavior.

derekparker commented on proc/threads_windows.go:
```diff
@@ -0,0 +1,171 @@
+		return false
+	}
+	switch fn.Name {
+	case "runtime.kevent", "runtime.mach_semaphore_wait", "runtime.usleep":
```
We can definitely remove "runtime.mach_semaphore_wait" from this list.

derekparker commented:
@lukehoban also, let's just go ahead and add the Appveyor badge to the README in the PR, since CI is already setup there it seems. That way it'll just be there when this lands.

lukehoban commented:
I've addressed all the code review feedback in the latest commits.

> #351 has landed, fixing the single step behavior.

Great.  I've updated the windows branch to align with this (actually just removing some code).

> let's just go ahead and add the Appveyor badge to the README in the PR, since CI is already setup there it seems.

I think you'll need to set up an AppVeyor project to track `derekparker/delve`.  The one I have at https://ci.appveyor.com/project/lukehoban/delve is pointing at `lukehoban/delve`.

derekparker commented:
Just merged this branch, will now start creating issues to track outstanding work that needs to be done, and I will get AppVeyor setup as well. Thank you @lukehoban for all your work on this!

lukehoban commented:
:zap:

kolkov commented:
This is very cool! 
@ignatov as you asked!

ahmetalpbalkan commented:
congrats!

techtonik commented:
@derekparker looks like this article https://blog.gopheracademy.com/advent-2015/debugging-with-delve/ should be updated with new info about Windows support. )
